bubble_sort.rb

```diff
@@ -0,0 +1,45 @@
+def bubble_sort(new_array)
+  flag = 0
+  while flag.zero?
+    flag = 1
+    new_array.each_with_index do |numb, position|
+      next if position == new_array.length - 1
+
+      test = yield(numb, new_array[position + 1])
+      next unless test.positive?
+
+      new_array[position] = new_array[position + 1]
+      new_array[position + 1] = numb
+      flag = 0
+    end
+  end
+end
+
+def bubble_sort_by(new_array)
+  flag = 0
+  while flag.zero?
+    flag = 1
+    new_array.each_with_index do |numb, position|
+      next if position == new_array.length - 1
+
+      test = yield(numb, new_array[position + 1])
+      next unless test.positive?
+
+      new_array[position] = new_array[position + 1]
+      new_array[position + 1] = numb
+      flag = 0
+    end
+  end
+end
```

Comment:
- Sort returns the sorted array. So your bubble_sort_by should do the same task. But it is not returning anything. That's why while i tried to execute the following code it did not work.
```ruby
print bubble_sort_by([1, 5, 3, 2, 1]) {|num1, num2| num1 - num2}
```
So, after line 32 you should place new_array.